Fix - Workloads requesting TAS cannot run via MultiKueue

IrvingMg opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind bug

#### What this PR does / why we need it:
It removes a validation that was preventing TAS from running with Multikueue.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #5313 

#### Special notes for your reviewer:
Added test files used for verifying this change on GKE.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
Fix: Removed a validation that was preventing TAS from running with MultiKueue.
```

IrvingMg commented:
TAS with Multikueue has been verified and is working as expected. Verification of TAS + DWS + Multikueue is currently in progress.

IrvingMg commented:
The issue `couldn't assign flavors to pod set main: no topology domains at level: kubernetes.io/hostname` seems to be fixed by adding an admission check with a provisioning request to the manager. 

However, when testing on GKE, the job is scheduled on the worker cluster, but the node is provisioned to the manager cluster. So, the pod scheduling fails with the error: Retrying after failure: Provisioning Request's pods cannot be scheduled in the node pool. Predicate checking errors: worker-us-east4-a-dws-nodepool (node(s) didn't match Pod's node affinity/selector).

IrvingMg commented:
As the expected setup involves configuring the provisioning request only in worker clusters, the following condition seems problematic: https://github.com/kubernetes-sigs/kueue/blob/main/pkg/scheduler/flavorassigner/tas_flavorassigner.go#L87. 

This is because it checks for a provisioning request to delay TAS, which causes failures in the manager cluster, resulting in the assignment being marked as inadmissible.

IrvingMg commented:
The condition mentioned above has been modified to allow delaying TAS when the ClusterQueue has MultiKueueAC, but it’s still not working. Based on the workload information shared below, it appears there is an issue with flavor assignment. The job is assigned to the `mgr-tas-flavor`, which is incorrect—it should be assigned to `wr-tas-flavor`. 

The setup used for testing can be found at: `site/static/examples/multikueue/gke/`.

Job description:
```
Name:         job-sample-job-tbhk9-f20d9
Namespace:    default
Labels:       kueue.x-k8s.io/job-uid=07c692f7-f97a-4fe6-aad2-5bca90ece1f8
Annotations:  <none>
API Version:  kueue.x-k8s.io/v1beta1
Kind:         Workload
Metadata:
  Creation Timestamp:  2025-06-18T17:15:34Z
  Finalizers:
    kueue.x-k8s.io/resource-in-use
  Generation:  1
  Owner References:
    API Version:           batch/v1
    Block Owner Deletion:  true
    Controller:            true
    Kind:                  Job
    Name:                  sample-job-tbhk9
    UID:                   07c692f7-f97a-4fe6-aad2-5bca90ece1f8
  Resource Version:        1750266938510031009
  UID:                     a0e3783e-317a-4022-849c-ab64943fdc89
Spec:
  Active:  true
  Pod Sets:
    Count:  1
    Name:   main
    Template:
      Metadata:
        Annotations:
          kueue.x-k8s.io/podset-required-topology:  kubernetes.io/hostname
      Spec:
        Containers:
          Args:
            10s
          Image:              gcr.io/k8s-staging-perf-tests/sleep:v0.0.3
          Image Pull Policy:  IfNotPresent
          Name:               dummy-job
          Resources:
            Limits:
              nvidia.com/gpu:  1
            Requests:
              Cpu:                     100m
              Memory:                  100Mi
              nvidia.com/gpu:          1
          Termination Message Path:    /dev/termination-log
          Termination Message Policy:  File
        Dns Policy:                    ClusterFirst
        Restart Policy:                Never
        Scheduler Name:                default-scheduler
        Security Context:
        Termination Grace Period Seconds:  30
        Tolerations:
          Effect:    NoSchedule
          Key:       nvidia.com/gpu
          Operator:  Exists
    Topology Request:
      Pod Index Label:    batch.kubernetes.io/job-completion-index
      Required:           kubernetes.io/hostname
  Priority:               0
  Priority Class Source:  
  Queue Name:             user-queue
Status:
  Admission:
    Cluster Queue:  mgr-cluster-queue
    Pod Set Assignments:
      Count:                     1
      Delayed Topology Request:  Pending
      Flavors:
        Cpu:             mgr-tas-flavor
        Memory:          mgr-tas-flavor
        nvidia.com/gpu:  mgr-tas-flavor
      Name:              main
      Resource Usage:
        Cpu:             100m
        Memory:          100Mi
        nvidia.com/gpu:  1
  Admission Checks:
    Last Transition Time:  2025-06-18T17:15:38Z
    Message:               The workload got reservation on "multikueue-test-worker-us"
    Name:                  multikueue-ac
    State:                 Pending
  Conditions:
    Last Transition Time:  2025-06-18T17:15:35Z
    Message:               Quota reserved in ClusterQueue mgr-cluster-queue
    Observed Generation:   1
    Reason:                QuotaReserved
    Status:                True
    Type:                  QuotaReserved
Events:
  Type    Reason         Age              From                       Message
  ----    ------         ----             ----                       -------
  Normal  QuotaReserved  8s               kueue-admission            Quota reserved in ClusterQueue mgr-cluster-queue, wait time since queued was 1s
  Normal  MultiKueue     5s (x4 over 6s)  kueue-workload-controller  The workload got reservation on "multikueue-test-worker-us"
```

IrvingMg commented:
Since tests with Multikueue and TAS are working fine, it's unlikely that there's a problem with flavor assignment.

IrvingMg commented:
New test shows that the job is admitted in the worker cluster and the node is provisioned correctly. However, the job is not starting, and as a result, the corresponding pods are not created. The job is unsuspended in the worker cluster but remains suspended in the manager cluster. Below are the object descriptions.

## Worker cluster

### Job
```bash
Name:             sample-job-glf8j
Namespace:        default
Selector:         batch.kubernetes.io/controller-uid=bd38d498-b9b1-46b5-a555-c3ba6e5b931a
Labels:           kueue.x-k8s.io/multikueue-origin=multikueue
                  kueue.x-k8s.io/prebuilt-workload-name=job-sample-job-glf8j-99cee
                  kueue.x-k8s.io/queue-name=user-queue
Annotations:      <none>
Parallelism:      1
Completions:      1
Completion Mode:  NonIndexed
Suspend:          false
Backoff Limit:    6
Pods Statuses:    0 Active / 0 Succeeded / 0 Failed
Pod Template:
  Labels:       batch.kubernetes.io/controller-uid=bd38d498-b9b1-46b5-a555-c3ba6e5b931a
                batch.kubernetes.io/job-name=sample-job-glf8j
                controller-uid=bd38d498-b9b1-46b5-a555-c3ba6e5b931a
                job-name=sample-job-glf8j
                kueue.x-k8s.io/podset=main
                kueue.x-k8s.io/tas=true
  Annotations:  autoscaling.x-k8s.io/consume-provisioning-request: job-sample-job-glf8j-99cee-dws-prov-1
                autoscaling.x-k8s.io/provisioning-class-name: queued-provisioning.gke.io
                cluster-autoscaler.kubernetes.io/consume-provisioning-request: job-sample-job-glf8j-99cee-dws-prov-1
                cluster-autoscaler.kubernetes.io/provisioning-class-name: queued-provisioning.gke.io
                kueue.x-k8s.io/podset-required-topology: kubernetes.io/hostname
                kueue.x-k8s.io/workload: job-sample-job-glf8j-99cee
  Containers:
   dummy-job:
    Image:      gcr.io/k8s-staging-perf-tests/sleep:v0.0.3
    Port:       <none>
    Host Port:  <none>
    Args:
      10s
    Limits:
      nvidia.com/gpu:  1
    Requests:
      cpu:             100m
      memory:          100Mi
      nvidia.com/gpu:  1
    Environment:       <none>
    Mounts:            <none>
  Volumes:             <none>
  Node-Selectors:      autoscaling.gke.io/provisioning-request=gke-default-job-sample-job-glf8j-51725ba0b10d35ce
                       cloud.google.com/gke-nodepool=worker-nodepool
  Tolerations:         cloud.google.com/gke-queued:NoSchedule op=Exists
                       nvidia.com/gpu:NoSchedule op=Exists
Events:
  Type    Reason                 Age    From                        Message
  ----    ------                 ----   ----                        -------
  Normal  UpdatedAdmissionCheck  4m20s  batch/job-kueue-controller  dws-prov: Waiting for resources. Currently there are not enough resources available to fulfill the request.
  Normal  Started                83s    batch/job-kueue-controller  Admitted by clusterQueue dws-cluster-queue
```

### Workload
```bash
Name:         job-sample-job-glf8j-99cee
Namespace:    default
Labels:       kueue.x-k8s.io/job-uid=bd38d498-b9b1-46b5-a555-c3ba6e5b931a
              kueue.x-k8s.io/multikueue-origin=multikueue
Annotations:  <none>
API Version:  kueue.x-k8s.io/v1beta1
Kind:         Workload
Metadata:
  Creation Timestamp:  2025-06-20T17:13:02Z
  Generation:          1
  Owner References:
    API Version:           batch/v1
    Block Owner Deletion:  true
    Controller:            true
    Kind:                  Job
    Name:                  sample-job-glf8j
    UID:                   bd38d498-b9b1-46b5-a555-c3ba6e5b931a
  Resource Version:        1750439784398767016
  UID:                     9bf0b05d-b363-4521-8256-8bc50b335e1b
Spec:
  Active:  true
  Pod Sets:
    Count:  1
    Name:   main
    Template:
      Metadata:
        Annotations:
          kueue.x-k8s.io/podset-required-topology:  kubernetes.io/hostname
      Spec:
        Containers:
          Args:
            10s
          Image:              gcr.io/k8s-staging-perf-tests/sleep:v0.0.3
          Image Pull Policy:  IfNotPresent
          Name:               dummy-job
          Resources:
            Limits:
              nvidia.com/gpu:  1
            Requests:
              Cpu:                     100m
              Memory:                  100Mi
              nvidia.com/gpu:          1
          Termination Message Path:    /dev/termination-log
          Termination Message Policy:  File
        Dns Policy:                    ClusterFirst
        Restart Policy:                Never
        Scheduler Name:                default-scheduler
        Security Context:
        Termination Grace Period Seconds:  30
        Tolerations:
          Effect:    NoSchedule
          Key:       nvidia.com/gpu
          Operator:  Exists
    Topology Request:
      Pod Index Label:    batch.kubernetes.io/job-completion-index
      Required:           kubernetes.io/hostname
  Priority:               0
  Priority Class Source:  
  Queue Name:             user-queue
Status:
  Admission:
    Cluster Queue:  dws-cluster-queue
    Pod Set Assignments:
      Count:                     1
      Delayed Topology Request:  Ready
      Flavors:
        Cpu:             wr-tas-flavor
        Memory:          wr-tas-flavor
        nvidia.com/gpu:  wr-tas-flavor
      Name:              main
      Resource Usage:
        Cpu:             100m
        Memory:          100Mi
        nvidia.com/gpu:  1
      Topology Assignment:
        Domains:
          Count:  1
          Values:
            gke-worker-us-east4-a-worker-nodepool-e22164d7-zdfb
        Levels:
          kubernetes.io/hostname
  Admission Checks:
    Last Transition Time:  2025-06-20T17:13:02Z
    Message:               Provisioning Request was successfully provisioned.
    Name:                  dws-prov
    Pod Set Updates:
      Annotations:
        autoscaling.x-k8s.io/consume-provisioning-request:              job-sample-job-glf8j-99cee-dws-prov-1
        autoscaling.x-k8s.io/provisioning-class-name:                   queued-provisioning.gke.io
        cluster-autoscaler.kubernetes.io/consume-provisioning-request:  job-sample-job-glf8j-99cee-dws-prov-1
        cluster-autoscaler.kubernetes.io/provisioning-class-name:       queued-provisioning.gke.io
      Name:                                                             main
      Node Selector:
        autoscaling.gke.io/provisioning-request:  gke-default-job-sample-job-glf8j-51725ba0b10d35ce
    State:                                        Ready
  Conditions:
    Last Transition Time:  2025-06-20T17:13:02Z
    Message:               Quota reserved in ClusterQueue dws-cluster-queue
    Observed Generation:   1
    Reason:                QuotaReserved
    Status:                True
    Type:                  QuotaReserved
    Last Transition Time:  2025-06-20T17:16:24Z
    Message:               The workload is admitted
    Observed Generation:   1
    Reason:                Admitted
    Status:                True
    Type:                  Admitted
Events:
  Type     Reason                      Age                     From                                   Message
  ----     ------                      ----                    ----                                   -------
  Normal   QuotaReserved               6m23s                   kueue-admission                        Quota reserved in ClusterQueue dws-cluster-queue, wait time since queued was 1s
  Normal   ProvisioningRequestCreated  6m23s                   kueue-provisioning-request-controller  Created ProvisioningRequest: "job-sample-job-glf8j-99cee-dws-prov-1"
  Warning  SecondPassFailed            3m53s (x13 over 4m5s)   kueue-admission                        couldn't assign flavors to pod set main: no topology domains at level: kubernetes.io/hostname
  Warning  SecondPassFailed            3m41s (x12 over 3m52s)  kueue-admission                        couldn't assign flavors to pod set main: topology "gke-default" doesn't allow to fit any of 1 pod(s)
  Normal   Admitted                    3m2s                    kueue-admission                        Admitted by ClusterQueue dws-cluster-queue, wait time since reservation was 0s
```

### ProvReq
```bash
Name:         job-sample-job-glf8j-99cee-dws-prov-1
Namespace:    default
Labels:       kueue.x-k8s.io/managed=true
Annotations:  <none>
API Version:  autoscaling.x-k8s.io/v1
Kind:         ProvisioningRequest
Metadata:
  Creation Timestamp:  2025-06-20T17:13:03Z
  Generation:          1
  Owner References:
    API Version:           kueue.x-k8s.io/v1beta1
    Block Owner Deletion:  true
    Controller:            true
    Kind:                  Workload
    Name:                  job-sample-job-glf8j-99cee
    UID:                   9bf0b05d-b363-4521-8256-8bc50b335e1b
  Resource Version:        1750439720047407005
  UID:                     90d52767-930f-47e3-8cd3-2af09e486fe5
Spec:
  Parameters:
  Pod Sets:
    Count:  1
    Pod Template Ref:
      Name:                 ppt-job-sample-job-glf8j-99cee-dws-prov-1-main
  Provisioning Class Name:  queued-provisioning.gke.io
Status:
  Conditions:
    Last Transition Time:  2025-06-20T17:13:26Z
    Message:               Provisioning Request was successfully queued.
    Observed Generation:   1
    Reason:                SuccessfullyQueued
    Status:                True
    Type:                  Accepted
    Last Transition Time:  2025-06-20T17:15:19Z
    Message:               Provisioning Request was successfully provisioned.
    Observed Generation:   1
    Reason:                Provisioned
    Status:                True
    Type:                  Provisioned
  Provisioning Class Details:
    Accelerator Type:            nvidia-tesla-t4
    Node Group Name:             gke-worker-us-east4-a-worker-nodepool-e22164d7-grp
    Node Pool Auto Provisioned:  false
    Node Pool Name:              worker-nodepool
    Resize Request Name:         gke-default-job-sample-job-glf8j-51725ba0b10d35ce
    Selected Zone:               us-east4-a
Events:                          <none>
```

## Manager

### Job
```bash
Name:             sample-job-glf8j
Namespace:        default
Selector:         batch.kubernetes.io/controller-uid=c54d47ae-e2bc-4d36-80d8-f4db37cd6874
Labels:           kueue.x-k8s.io/queue-name=user-queue
Annotations:      <none>
Parallelism:      1
Completions:      1
Completion Mode:  NonIndexed
Suspend:          true
Backoff Limit:    6
Pods Statuses:    0 Active / 0 Succeeded / 0 Failed
Pod Template:
  Labels:       batch.kubernetes.io/controller-uid=c54d47ae-e2bc-4d36-80d8-f4db37cd6874
                batch.kubernetes.io/job-name=sample-job-glf8j
                controller-uid=c54d47ae-e2bc-4d36-80d8-f4db37cd6874
                job-name=sample-job-glf8j
  Annotations:  kueue.x-k8s.io/podset-required-topology: kubernetes.io/hostname
  Containers:
   dummy-job:
    Image:      gcr.io/k8s-staging-perf-tests/sleep:v0.0.3
    Port:       <none>
    Host Port:  <none>
    Args:
      10s
    Limits:
      nvidia.com/gpu:  1
    Requests:
      cpu:             100m
      memory:          100Mi
      nvidia.com/gpu:  1
    Environment:       <none>
    Mounts:            <none>
  Volumes:             <none>
  Node-Selectors:      <none>
  Tolerations:         nvidia.com/gpu:NoSchedule op=Exists
Events:
  Type    Reason                 Age                    From                        Message
  ----    ------                 ----                   ----                        -------
  Normal  CreatedWorkload        9m33s                  batch/job-kueue-controller  Created Workload: default/job-sample-job-glf8j-99cee
  Normal  UpdatedAdmissionCheck  6m10s (x6 over 9m31s)  batch/job-kueue-controller  multikueue-ac: The workload got reservation on "multikueue-test-worker-us"
```

### Workload
```bash
Name:         job-sample-job-glf8j-99cee
Namespace:    default
Labels:       kueue.x-k8s.io/job-uid=c54d47ae-e2bc-4d36-80d8-f4db37cd6874
Annotations:  <none>
API Version:  kueue.x-k8s.io/v1beta1
Kind:         Workload
Metadata:
  Creation Timestamp:  2025-06-20T17:13:02Z
  Finalizers:
    kueue.x-k8s.io/resource-in-use
  Generation:  1
  Owner References:
    API Version:           batch/v1
    Block Owner Deletion:  true
    Controller:            true
    Kind:                  Job
    Name:                  sample-job-glf8j
    UID:                   c54d47ae-e2bc-4d36-80d8-f4db37cd6874
  Resource Version:        1750439785725023016
  UID:                     91898efd-37ef-40b1-9953-500de5bc0b6d
Spec:
  Active:  true
  Pod Sets:
    Count:  1
    Name:   main
    Template:
      Metadata:
        Annotations:
          kueue.x-k8s.io/podset-required-topology:  kubernetes.io/hostname
      Spec:
        Containers:
          Args:
            10s
          Image:              gcr.io/k8s-staging-perf-tests/sleep:v0.0.3
          Image Pull Policy:  IfNotPresent
          Name:               dummy-job
          Resources:
            Limits:
              nvidia.com/gpu:  1
            Requests:
              Cpu:                     100m
              Memory:                  100Mi
              nvidia.com/gpu:          1
          Termination Message Path:    /dev/termination-log
          Termination Message Policy:  File
        Dns Policy:                    ClusterFirst
        Restart Policy:                Never
        Scheduler Name:                default-scheduler
        Security Context:
        Termination Grace Period Seconds:  30
        Tolerations:
          Effect:    NoSchedule
          Key:       nvidia.com/gpu
          Operator:  Exists
    Topology Request:
      Pod Index Label:    batch.kubernetes.io/job-completion-index
      Required:           kubernetes.io/hostname
  Priority:               0
  Priority Class Source:  
  Queue Name:             user-queue
Status:
  Admission:
    Cluster Queue:  mgr-cluster-queue
    Pod Set Assignments:
      Count:                     1
      Delayed Topology Request:  Pending
      Flavors:
        Cpu:             mgr-tas-flavor
        Memory:          mgr-tas-flavor
        nvidia.com/gpu:  mgr-tas-flavor
      Name:              main
      Resource Usage:
        Cpu:             100m
        Memory:          100Mi
        nvidia.com/gpu:  1
  Admission Checks:
    Last Transition Time:  2025-06-20T17:16:25Z
    Message:               The workload got reservation on "multikueue-test-worker-us"
    Name:                  multikueue-ac
    State:                 Pending
  Conditions:
    Last Transition Time:  2025-06-20T17:13:02Z
    Message:               Quota reserved in ClusterQueue mgr-cluster-queue
    Observed Generation:   1
    Reason:                QuotaReserved
    Status:                True
    Type:                  QuotaReserved
Events:
  Type    Reason         Age                   From                       Message
  ----    ------         ----                  ----                       -------
  Normal  QuotaReserved  10m                   kueue-admission            Quota reserved in ClusterQueue mgr-cluster-queue, wait time since queued was 0s
  Normal  MultiKueue     7m23s (x11 over 10m)  kueue-workload-controller  The workload got reservation on "multikueue-test-worker-us"
```

IrvingMg commented:
> @IrvingMg does TAS already support MultiKueue in this branch (without ProvReq)?
> 
> It could be valuable as the first step.

Unfortunately, not yet. I just double-checked on GKE, and the behavior is the same as with MK + TAS + ProvReq. The workload gets a reservation and is admitted on the worker cluster, but for some reason, the pods aren't being created.

The integration tests added to this branch seem to pass because they only verify that the workload is admitted, which does happen on GKE as well.

IrvingMg commented:
## Task Summary (as of August 20)

### Goal
Enable MultiKueue (MK), Topology Aware Scheduling (TAS), and ProvisioningRequest to work correctly together on GKE with Dynamic Workload Shaping (DWS) enabled.

---

### What’s Been Done

1. Removed the restriction that explicitly blocked the use of MultiKueue with TAS in [`clusterqueue.go`](https://github.com/kubernetes-sigs/kueue/blob/8def1eef499bd463eb8a3a6113a44059bf3b1aa0/pkg/cache/clusterqueue.go#L281-L284).

2. After this change, encountered the error:  
   ```
   couldn't assign flavors to pod set main: no topology domains at level: kubernetes.io/hostname
   ```

3. Added integration tests in `test/integration/multikueue/tas/suite_test.go` for:
   - MK + TAS
   - MK + TAS + ProvisioningRequest

   These tests passed, but they only verify that the workload is admitted to the correct worker cluster, which also happens correctly in real GKE testing.

4. Identified an issue in MK setups where topology requests were not delayed as expected when DWS is enabled.

   In DWS-enabled scenarios, the scheduler is expected to delay topology assignment until the ProvisioningRequest is fulfilled (e.g., when a new node becomes available). However, in MK setups, the scheduler tried to assign topology too early, before the needed resources were ready. 

   This was caused by a condition in [`tas_flavorassigner.go`](https://github.com/kubernetes-sigs/kueue/blob/8def1eef499bd463eb8a3a6113a44059bf3b1aa0/pkg/scheduler/flavorassigner/tas_flavorassigner.go#L87), which required the workload to have both a `QuotaReservation` and a `ProvisioningRequest`.

   That condition is never satisfied in the manager cluster, since the scheduler runs there and its ClusterQueues do not include ProvisioningRequest admission checks. As a result, the topology assignment was not delayed, causing workloads to be marked inadmissible with the error above.

5. Fixed the condition by changing the logic to check whether the ClusterQueue has the MultiKueue admission check instead.

   The updated logic assumes that if a workload has a ProvisioningRequest and the ClusterQueue does not have the MK admission check, it is in a worker cluster, and topology assignment should be delayed there.


---

### Current Behavior (MK + TAS + ProvReq)
- When a job is created, the workload is `QuotaReserved` in both the manager and worker clusters.  
- The worker cluster creates a ProvisioningRequest and waits for resources to be provisioned.  
- Once the resources are available, the workload is admitted and runs (my test job executes a simple sleep command, so I assume it is running).
- After the workload completes, both the workload and the job are cleaned up in the worker cluster.  
- In the manager cluster, the workload is marked as `Finished`.  
- However, the job in the manager cluster remains stuck in the `Running` state.  
- This behavior is the same as observed with MK + TAS alone, suggesting a general issue with job status synchronization in MultiKueue setups.

---

### Environment & Testing
Testing was done on a real GKE cluster using Terraform scripts and YAML files (all included in the current PR). The setup covers MultiKueue, TAS, and ProvisioningRequest, and also includes files for testing TAS + ProvisioningRequest in a single-cluster environment, which was verified to work correctly on GKE.

A README file is also included to help understand and reproduce the cluster setup.

The code changes described above are also part of this current PR.

IrvingMg commented:
> * In addition to the sync issue, even though the workload appears to run correctly in the worker cluster, events in the workload object on the manager cluster show repeated attempts to admit it there, each failing with the known error:  `couldn't assign flavors to pod set main: no topology domains at level: kubernetes.io/hostname`.
>   This should not happen in the manager cluster. Topology Aware Scheduling should be handled in the worker cluster. These repeated TAS failures in the manager suggest that TAS may not be fully aware of MultiKueue, and could be interfering with proper job status synchronization from worker to manager.

It seems that simply disabling Topology Aware Scheduling in the manager cluster avoids the issue, and everything else works as expected. I've tested this with MultiKueue and TAS.

I'll continue testing with ProvisioningRequest and also work on adding e2e tests.

/cc @mbobrovskyi

IrvingMg commented:
After testing different scenarios with TAS disabled in the manager cluster, I believe the problem described above was a configuration issue, not a code issue. The integration of MultiKueue + TAS + ProvisioningRequest was initially tested under the assumption that TAS needed to be enabled in both the manager and worker clusters.
 
However, enabling TAS in the manager cluster makes it try to schedule based on its own local topology, which isn’t the intended behavior, since workloads are meant to execute in the worker clusters, not in the manager. Because the manager lacks access to topology data from the workers, this leads to scheduling failures.
 
Disabling TAS in the manager correctly delegates Topology Aware Scheduling to the worker clusters, where workloads are actually scheduled and run.
 
I also plan to revert the previous code changes made in this PR, which were intended to support TAS in the manager with MultiKueue. That approach added unnecessary complexity, which isn’t needed with the correct configuration in place.
 
I'll also add e2e tests to validate the correct behavior with this setup.

tenzen-y commented:
> After testing different scenarios with TAS disabled in the manager cluster, I believe the problem described above was a configuration issue, not a code issue. The integration of MultiKueue + TAS + ProvisioningRequest was initially tested under the assumption that TAS needed to be enabled in both the manager and worker clusters.
> 
> However, enabling TAS in the manager cluster makes it try to schedule based on its own local topology, which isn’t the intended behavior, since workloads are meant to execute in the worker clusters, not in the manager. Because the manager lacks access to topology data from the workers, this leads to scheduling failures.
> 
> Disabling TAS in the manager correctly delegates Topology Aware Scheduling to the worker clusters, where workloads are actually scheduled and run.
> 
> I also plan to revert the previous code changes made in this PR, which were intended to support TAS in the manager with MultiKueue. That approach added unnecessary complexity, which isn’t needed with the correct configuration in place.
> 
> I'll also add e2e tests to validate the correct behavior with this setup.

Thank you for investigating that. 

> After testing different scenarios with TAS disabled in the manager cluster

Does this mean not deploying Topology resource in manager cluster? Or removing TAS annotations from workloads?

BTW, I'm looking into cluster role sharing, which a single cluster is responsible for both manager and worker: https://github.com/kubernetes-sigs/kueue/issues/5704

So, supporting TAS both in the manager and the worker would be nice. But, the supporting TAS with cluster sharing could be considered as another feature. So, I'm ok with narrow down this PR scope not to consider cluster sharing environment.

tenzen-y commented:
> I agree that this might be a blocker for TAS graduation. In the current implementation, enabling TAS without topology resources in the manager results in the following error:
> 
> couldn't assign flavors to pod set main: Flavor 'mgr-tas-flavor' does not support TopologyAwareScheduling
> 
> And when topologies are added, the ClusterQueue becomes inactive with the message:
> 
> Can't admit new workloads: TAS is not supported with MultiKueue admission check
> 
> So, simply enabling the TAS feature gate in the manager is currently causing issues when creating a job that requires topology.

Yes, we have validations in https://github.com/kubernetes-sigs/kueue/blob/99b9a12b5444342bffff5d594b11ff66d148b481/pkg/cache/clusterqueue.go#L280-L291.
What if we remove this validation, then add topology reference in all resource flavors?

tenzen-y commented:
> I believe we might need to remove the validation here, but move it elsewhere to soften the impact, as making the entire ClusterQueue inactive seems too aggressive.

I think that validation makes reasonable to me since this is an effective way to inform incompatible feature combinations of TAS and MuliKueue to cluster admins.

> For example, we could check this condition only when attempting to schedule a workload that requests a topology, and then fail or ignore only that workload if the condition isn't met, instead of deactivating the whole ClusterQueue.

The problem is TAS implicit mode. The implicit mode automatically calculates Topology Placement even when the users do not request TAS, explicitly.

mimowo commented:
I think ideally we should have a solution which allows to enable TAS  on the management cluster as this is soon to be default (when we graduate to Beta). 

I guess this may require taking a step back and designing the proper integration. cc @mwysokin

IrvingMg commented:
I've added integration and e2e tests. Integration tests are working as expected for Multikueue, TAS, and Provisioning Request. I've also added e2e tests for Multikueue and TAS, but encountered some issues that I'm currently investigating.

IrvingMg commented:
/test all

IrvingMg commented:
/test all

IrvingMg commented on pkg/controller/admissionchecks/multikueue/indexer.go:
```diff
@@ -77,7 +77,8 @@ func SetupIndexer(ctx context.Context, indexer client.FieldIndexer, configNamesp
+	// WorkloadsWithAdmissionCheckKey may already be registered by provisioning controller
+	if err := indexer.IndexField(ctx, &kueue.Workload{}, WorkloadsWithAdmissionCheckKey, indexWorkloadsAdmissionChecks); err != nil && !admissioncheck.IsIndexerConflict(err) {
 		return fmt.Errorf("setting index on workloads admission checks: %w", err)
 	}
```
Open #7432

IrvingMg commented on pkg/controller/admissionchecks/multikueue/workload.go:
```diff
@@ -420,6 +421,18 @@ func (w *wlReconciler) reconcileGroup(ctx context.Context, group *wlGroup) (reco
 			return reconcile.Result{}, err
 		}
 
+		needsTopologySync := needsTopologySyncFromRemote(group.local, group.remotes[reservingRemote])
```
I've updated this so that we now sync only the topology readiness state via DelayedTopologyRequest. This allows the manager workload to be fully admitted while keeping the actual topology assignment private to the worker cluster.

IrvingMg replied:
See [sample-job-446jc.txt](https://github.com/user-attachments/files/23295093/sample-job-446jc.txt)  for an example tracking log that demonstrates the full end-to-end workflow.

mimowo commented on pkg/util/admissioncheck/admissioncheck.go:
```diff
@@ -234,3 +234,9 @@ func GetRemoteClusters(ctx context.Context, helper *MultiKueueStoreHelper, acNam
+
+// IsIndexerConflict returns true if the error is an indexer conflict error.
+// This is used to detect when multiple controllers try to register the same index.
+func IsIndexerConflict(err error) bool {
```
Rebase to revert this change

mimowo commented on pkg/controller/admissionchecks/multikueue/workload.go:
```diff
@@ -637,6 +618,98 @@ func (w *wlReconciler) setupWithManager(mgr ctrl.Manager) error {
+		for j := range local.Status.Admission.PodSetAssignments {
+			if local.Status.Admission.PodSetAssignments[j].Name != remotePSA.Name {
+				continue
+			}
```
Please commonize the code for getting the PodSetAssignment by name and reuse the same helper in here: https://github.com/kubernetes-sigs/kueue/pull/5361/files#r2486006884

mimowo commented on pkg/controller/admissionchecks/multikueue/workload.go:
```diff
@@ -637,6 +618,98 @@ func (w *wlReconciler) setupWithManager(mgr ctrl.Manager) error {
+	return nil
+}
+
+func updateDelayedTopologyRequest(local, remote *kueue.Workload) {
```
This looks very similar to `needsDelayedTopologyUpdate`. Think more how to commonize the code

IrvingMg commented:
/test pull-kueue-test-e2e-multikueue-main

Due to #7457
